src/libraries/System.Private.CoreLib/src/System/Reflection/Nullability/NullabilityInfoContext.cs

```diff
@@ -0,0 +1,430 @@
+{
+    public sealed class NullabilityInfoContext
+    {
+        private const string CompilerServicesNameSpace = "System.Runtime.CompilerServices";
+        private readonly Dictionary<Module, NotAnnotatedStatus> _publicOnlyModules = new();
+        private readonly Dictionary<MemberInfo, NullabilityState> _context = new();
+        private static readonly NullabilityInfo[] s_emptyArray = Array.Empty<NullabilityInfo>();
+
+        [Flags]
+        private enum NotAnnotatedStatus
+        {
+            None = 0x0, // no restriction, all members annotated
+            Private = 0x1, // private members not annotated
+            Internal = 0x2 // internal members not annotated
+        }
+
+        private NullabilityState GetNullableContext(MemberInfo? memberInfo)
+        {
+            while (memberInfo != null)
+            {
+                if (_context.TryGetValue(memberInfo, out var state))
+                {
+                    return state;
+                }
+
+                var attributes = memberInfo.GetCustomAttributesData();
+                foreach (var attribute in attributes)
+                {
+                    if (attribute.AttributeType.Name == "NullableContextAttribute" &&
+                        attribute.AttributeType.Namespace == CompilerServicesNameSpace &&
+                        attribute.ConstructorArguments.Count == 1)
+                    {
+                        state = TranslateByte(attribute.ConstructorArguments[0].Value);
+                        _context.Add(memberInfo, state);
+                        return state;
+                    }
+                }
+
+                memberInfo = memberInfo.DeclaringType;
+            }
+
+            return NullabilityState.Unknown;
+        }
+
+        [RequiresUnreferencedCode("Nullability attributes are trimmed by the linker")]
+        public NullabilityInfo Create(ParameterInfo parameterInfo)
+        {
+            if (parameterInfo.Member is MethodBase method &&
+                (method.IsPrivate || method.IsFamilyAndAssembly || method.IsAssembly))
+            {
+                if (IsPublicOnly(method.IsPrivate, method.IsFamilyAndAssembly, method.IsAssembly, method.Module))
+                {
+                    return new NullabilityInfo(parameterInfo.ParameterType, NullabilityState.Unknown, NullabilityState.Unknown, null, s_emptyArray);
+                }
+            }
+
+            var attributes = parameterInfo.GetCustomAttributesData();
+            var nullability = GetNullabilityInfo(parameterInfo.Member, parameterInfo.ParameterType, attributes);
+
+            if (nullability.ReadState != NullabilityState.Unknown)
+            {
+                CheckParameterMetadataType(parameterInfo, nullability);
+            }
+
+            CheckNullabilityAttributes(nullability, attributes);
+            return nullability;
+        }
+
+        [RequiresUnreferencedCode("Nullability attributes are trimmed by the linker")]
```

Comment:
Getting Trim analysis error IL2026 for `Module.ResolveMethod(MetadataToken)` usage, the nullability attributes are not available for trimmed builds anyway, so added the same attribute to get rid of the warning and also to warn for trimmed apps CC @eerhardt

Reply:
Can you explain why you need to write code like:

```
var metaMethod = (MethodInfo)method.Module.ResolveMethod(method.MetadataToken)!;
```

How/why are `metaMethod` and `method` different? Maybe a comment in the code would help.

In all reality, I believe you can just suppress the IL2026 warning for `Module.ResolveMethod` for the same reasoning as here:

https://github.com/dotnet/runtime/blob/779b79401648908279d0903b5a91f593c745efd0/src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/ModuleBuilder.cs#L431-L436

Reply:
If you need to get generic method definition, reflection has dedicated methods for that, like `MethodInfo.GetGenericMethodDefinition`. These dedicated methods are both linker friendly and faster compared roundtripping via method token.

Reply:
> How/why are `metaMethod` and `method` different? Maybe a comment in the code would help.

They will be different if the generic type used as a parameter or return value. For example for `List<string?> list` the method will look like `list.Add(string item)` and the item would be evaluated to not null when it supposed to be nullable, the metaMethod would look like `list.Add(T item)` and the item nullability will be nullable as expected.  

Note that it will be the same for `List<string> list`, there is no information about that concrete type created with `?` or not

> In all reality, I believe you can just suppress the IL2026 warning for Module.ResolveMethod for the same reasoning as here:

I can suppress it if we do not want to warn that it will not work for trimmed assembly, probably that is better as if the nullability attributes missing the API would just return `NullableState.Unknown`

Reply:
> if we do not want to warn that it will not work for trimmed assembly

See my other [comment](https://github.com/dotnet/runtime/pull/54985#discussion_r662400510) about the overall behavior here for a trimmed app.

Reply:
> If you need to get generic method definition, reflection has dedicated methods for that, like `MethodInfo.GetGenericMethodDefinition`. These dedicated methods are both linker friendly and faster compared roundtripping via method token.

`MethodInfo.GetGenericMethodDefinition` is working only for methods with generic type parameters, it's throwing for other cases like trying to load meta-method for `List.Add(T item)`

Reply:
> See my other comment about the overall behavior here for a trimmed app.

I leave it as is until that question answered

Reply:
Try this:
```
    static MethodInfo StripInstantiation(MethodInfo m)
    {
        if (m.IsGenericMethod && !m.IsGenericMethodDefinition)
            m = m.GetGenericMethodDefinition();
        Type? t = m.DeclaringType;
        if ((t != null) && t.IsGenericType && !t.IsGenericTypeDefinition)
            m = (MethodInfo)t.GetGenericTypeDefinition().GetMemberWithSameMetadataDefinitionAs(m);
        return m;
    }
```

Reply:
> Try this:

That worked well, thanks, I have updated accordingly methods for loading metaType for FieldInfo, PropertyInfo too, now the `RequiresUnreferencedCode` attributes are removed as not using `Module.ResolveMember(MetadataToken)` and no linker warning anymore